File: android-app/app/src/main/java/com/lasohealth/android/core/analysis/analyzers/NutritionCorrelationAnalyzer.kt
package com.lasohealth.android.core.analysis.analyzers

import com.lasohealth.android.core.analysis.AnalysisContext
import com.lasohealth.android.core.analysis.AnalysisInsight
import com.lasohealth.android.core.analysis.DailySample
import com.lasohealth.android.core.analysis.InsightAnalyzer
import com.lasohealth.android.core.analysis.InsightDirective
import com.lasohealth.android.core.analysis.InsightPriority
import com.lasohealth.android.core.model.HealthMetric
import java.util.Calendar
import java.util.UUID
import kotlin.math.abs
import kotlin.math.sqrt

/**
 * Time-aware correlation engine for nutrition -> health outcome pairs.
 * Tests ~16 curated physiologically plausible pairs (not blind N^2 search).
 *
 * Ported from iOS NutritionCorrelationAnalyzer.swift.
 */
class NutritionCorrelationAnalyzer : InsightAnalyzer {

    override val name: String = "NutritionCorrelationAnalyzer"

    override suspend fun generateInsights(context: AnalysisContext): List<AnalysisInsight> {
        val correlations = analyze(context.timeSeries)
        return generateInsightsFromCorrelations(correlations)
    }

    // ----- Curated Pairs -----

    private data class CuratedPair(
        val nutrition: HealthMetric,
        val outcome: HealthMetric,
        val dayOffset: Int,
    )

    private val curatedPairs = listOf(
        // Caffeine -> sleep
        CuratedPair(HealthMetric.CAFFEINE_INTAKE, HealthMetric.SLEEP_DURATION, 0),
        CuratedPair(HealthMetric.CAFFEINE_INTAKE, HealthMetric.SLEEP_DEEP, 0),
        CuratedPair(HealthMetric.CAFFEINE_INTAKE, HealthMetric.SLEEP_REM, 0),
        CuratedPair(HealthMetric.CAFFEINE_INTAKE, HealthMetric.RESTING_HEART_RATE, 0),
        CuratedPair(HealthMetric.CAFFEINE_INTAKE, HealthMetric.HEART_RATE_VARIABILITY, 1),
        // Hydration -> outcomes
        CuratedPair(HealthMetric.WATER_INTAKE, HealthMetric.HEART_RATE_VARIABILITY, 1),
        CuratedPair(HealthMetric.WATER_INTAKE, HealthMetric.RESTING_HEART_RATE, 1),
        // Protein -> recovery
        CuratedPair(HealthMetric.PROTEIN_INTAKE, HealthMetric.HEART_RATE_VARIABILITY, 1),
        // Sugar -> sleep
        CuratedPair(HealthMetric.SUGAR_INTAKE, HealthMetric.SLEEP_DEEP, 0),
        CuratedPair(HealthMetric.SUGAR_INTAKE, HealthMetric.HEART_RATE_VARIABILITY, 1),
        // Sodium -> BP
        CuratedPair(HealthMetric.SODIUM_INTAKE, HealthMetric.BLOOD_PRESSURE_SYSTOLIC, 1),
        // Fiber -> glucose stability
        CuratedPair(HealthMetric.FIBER_INTAKE, HealthMetric.BLOOD_GLUCOSE, 0),
        // Total calories -> activity
        CuratedPair(HealthMetric.TOTAL_CALORIES_INTAKE, HealthMetric.ACTIVE_CALORIES, 1),
        // Fat -> sleep
        CuratedPair(HealthMetric.FAT_INTAKE, HealthMetric.SLEEP_DEEP, 0),
        // Carbs -> activity
        CuratedPair(HealthMetric.CARBOHYDRATE_INTAKE, HealthMetric.EXERCISE_MINUTES, 1),
    )

    // ----- Analysis -----

    private data class NutritionCorrelation(
        val nutritionMetric: HealthMetric,
        val outcomeMetric: HealthMetric,
        val correlation: Double,
        val thresholdValue: Double?,
        val thresholdImpact: Double?,
        val sampleCount: Int,
        val insightText: String,
    )

    private fun analyze(timeSeries: Map<HealthMetric, List<DailySample>>): List<NutritionCorrelation> {
        val results = mutableListOf<NutritionCorrelation>()

        for (pair in curatedPairs) {
            val nutritionSamples = timeSeries[pair.nutrition] ?: continue
            val outcomeSamples = timeSeries[pair.outcome] ?: continue
            if (nutritionSamples.size < 14 || outcomeSamples.size < 14) continue

            val nutritionByDate = buildDateLookup(nutritionSamples)
            val outcomeByDate = buildDateLookup(outcomeSamples)

            val nutritionValues = mutableListOf<Double>()
            val outcomeValues = mutableListOf<Double>()

            for ((date, nv) in nutritionByDate) {
                val targetDate = date + pair.dayOffset.toLong() * DAY_MS
                val ov = outcomeByDate[targetDate] ?: continue
                nutritionValues += nv
                outcomeValues += ov
            }

            if (nutritionValues.size < 10) continue

            val r = pearsonCorrelation(nutritionValues, outcomeValues) ?: continue
            if (abs(r) < 0.15) continue

            val threshold = detectThreshold(nutritionValues, outcomeValues)
            val insightText = generateInsightText(pair, r, threshold, nutritionValues.size)

            results += NutritionCorrelation(
                nutritionMetric = pair.nutrition,
                outcomeMetric = pair.outcome,
                correlation = r,
                thresholdValue = threshold?.first,
                thresholdImpact = threshold?.second,
                sampleCount = nutritionValues.size,
                insightText = insightText,
            )
        }

        return results.sortedByDescending { abs(it.correlation) }
    }

    private fun generateInsightsFromCorrelations(correlations: List<NutritionCorrelation>): List<AnalysisInsight> {
        return correlations.take(5).map { corr ->
            AnalysisInsight(
                id = UUID.randomUUID().toString(),
                title = "${corr.nutritionMetric.displayName} Affects ${corr.outcomeMetric.displayName}",
                detail = corr.insightText,
                metric = corr.nutritionMetric,
                priority = if (abs(corr.correlation) > 0.4) InsightPriority.MEDIUM else InsightPriority.INFORMATIONAL,
                directive = InsightDirective.INFORMATIONAL,
                confidence = minOf(corr.sampleCount / 60.0, 1.0),
            )
        }
    }

    // ----- Threshold Detection -----

    private fun detectThreshold(
        nutritionValues: List<Double>,
        outcomeValues: List<Double>,
    ): Pair<Double, Double>? {
        val sorted = nutritionValues.zip(outcomeValues).sortedBy { it.first }
        val n = sorted.size
        if (n < 20) return null

        val q3Start = n * 3 / 4
        val bottomHalf = sorted.take(n / 2).map { it.second }
        val topQuartile = sorted.drop(q3Start).map { it.second }

        if (bottomHalf.isEmpty() || topQuartile.isEmpty()) return null

        val bottomAvg = bottomHalf.average()
        val topAvg = topQuartile.average()
        if (bottomAvg == 0.0) return null

        val impact = ((topAvg - bottomAvg) / abs(bottomAvg)) * 100.0
        if (abs(impact) < 5.0) return null

        val thresholdValue = sorted[q3Start].first
        return thresholdValue to impact
    }

    // ----- Text Generation -----

    private fun generateInsightText(
        pair: CuratedPair,
        correlation: Double,
        threshold: Pair<Double, Double>?,
        sampleCount: Int,
    ): String {
        val direction = if (correlation > 0) "increases" else "decreases"
        val strength = when {
            abs(correlation) > 0.5 -> "strongly"
            abs(correlation) > 0.3 -> "moderately"
            else -> "slightly"
        }

        var text = "Your ${pair.nutrition.displayName} $strength $direction ${pair.outcome.displayName}"
        if (pair.dayOffset > 0) text += " the next day"
        text += " (r=${String.format("%.2f", correlation)}, n=$sampleCount)."

        threshold?.let { (value, impact) ->
            text += " Impact becomes more pronounced above ${String.format("%.0f", value)} ${pair.nutrition.unit} " +
                "(${String.format("%.0f", abs(impact))}% change)."
        }

        return text
    }

    // ----- Helpers -----

    private fun buildDateLookup(samples: List<DailySample>): Map<Long, Double> {
        val map = mutableMapOf<Long, Double>()
        for (s in samples) map[startOfDay(s.date)] = s.value
        return map
    }

    private fun pearsonCorrelation(x: List<Double>, y: List<Double>): Double? {
        val n = x.size.toDouble()
        if (n <= 2) return null
        val sumX = x.sum(); val sumY = y.sum()
        val sumXY = x.zip(y).sumOf { it.first * it.second }
        val sumX2 = x.sumOf { it * it }; val sumY2 = y.sumOf { it * it }

        val num = n * sumXY - sumX * sumY
        val den = sqrt((n * sumX2 - sumX * sumX) * (n * sumY2 - sumY * sumY))
        return if (den > 0) num / den else null
    }

    private fun startOfDay(epochMs: Long): Long {
        val cal = Calendar.getInstance().apply { timeInMillis = epochMs }
        cal.set(Calendar.HOUR_OF_DAY, 0); cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0); cal.set(Calendar.MILLISECOND, 0)
        return cal.timeInMillis
    }

    companion object {
        private const val DAY_MS = 86_400_000L
    }
}

File: Core/Analysis/NutritionCorrelationAnalyzer.swift
import Foundation

/// Time-aware correlation engine for nutrition → health outcome pairs.
/// Tests ~16 curated physiologically plausible pairs (not blind N² search).
struct NutritionCorrelationAnalyzer {

    // MARK: - Types

    struct NutritionCorrelation: Identifiable {
        let id = UUID()
        let nutritionMetric: HealthMetric
        let outcomeMetric: HealthMetric
        let correlation: Double          // Pearson r
        let timingEffect: TimingEffect?  // For timing-sensitive pairs
        let thresholdValue: Double?      // Threshold where impact becomes pronounced
        let actionableInsight: String
        let sampleCount: Int
    }

    struct TimingEffect {
        let cutoffHour: Int              // e.g., 14 for 2pm
        let beforeCutoffCorrelation: Double
        let afterCutoffCorrelation: Double
        let isTimingSensitive: Bool
    }

    // MARK: - Curated Pairs

    private struct CuratedPair {
        let nutrition: HealthMetric
        let outcome: HealthMetric
        let dayOffset: Int                // 0 = same day, 1 = next day
        let timingSensitive: Bool
        let cutoffHour: Int?              // For timing-sensitive pairs
    }

    private static let curatedPairs: [CuratedPair] = [
        // Caffeine → sleep (timing-sensitive, cutoff 2pm)
        CuratedPair(nutrition: .caffeineIntake, outcome: .sleepDuration, dayOffset: 0, timingSensitive: true, cutoffHour: 14),
        CuratedPair(nutrition: .caffeineIntake, outcome: .sleepDeep, dayOffset: 0, timingSensitive: true, cutoffHour: 14),
        CuratedPair(nutrition: .caffeineIntake, outcome: .sleepREM, dayOffset: 0, timingSensitive: true, cutoffHour: 14),
        CuratedPair(nutrition: .caffeineIntake, outcome: .restingHeartRate, dayOffset: 0, timingSensitive: false, cutoffHour: nil),
        CuratedPair(nutrition: .caffeineIntake, outcome: .heartRateVariability, dayOffset: 1, timingSensitive: false, cutoffHour: nil),

        // Hydration → outcomes
        CuratedPair(nutrition: .waterIntake, outcome: .heartRateVariability, dayOffset: 1, timingSensitive: false, cutoffHour: nil),
        CuratedPair(nutrition: .waterIntake, outcome: .restingHeartRate, dayOffset: 1, timingSensitive: false, cutoffHour: nil),

        // Protein → recovery
        CuratedPair(nutrition: .proteinIntake, outcome: .heartRateVariability, dayOffset: 1, timingSensitive: false, cutoffHour: nil),

        // Sugar → sleep (timing-sensitive, cutoff 6pm)
        CuratedPair(nutrition: .sugarIntake, outcome: .sleepDeep, dayOffset: 0, timingSensitive: true, cutoffHour: 18),
        CuratedPair(nutrition: .sugarIntake, outcome: .heartRateVariability, dayOffset: 1, timingSensitive: false, cutoffHour: nil),

        // Sodium → BP
        CuratedPair(nutrition: .sodiumIntake, outcome: .bloodPressureSystolic, dayOffset: 1, timingSensitive: false, cutoffHour: nil),

        // Fiber → glucose stability
        CuratedPair(nutrition: .fiberIntake, outcome: .bloodGlucose, dayOffset: 0, timingSensitive: false, cutoffHour: nil),

        // Total calories → activity
        CuratedPair(nutrition: .totalCaloriesIntake, outcome: .activeCalories, dayOffset: 1, timingSensitive: false, cutoffHour: nil),

        // Fat → sleep
        CuratedPair(nutrition: .fatIntake, outcome: .sleepDeep, dayOffset: 0, timingSensitive: false, cutoffHour: nil),

        // Carbs → activity
        CuratedPair(nutrition: .carbohydrateIntake, outcome: .exerciseMinutes, dayOffset: 1, timingSensitive: false, cutoffHour: nil),
    ]

    // MARK: - Analysis

    /// Analyze curated nutrition-outcome pairs using available time series data.
    static func analyze(timeSeries: [HealthMetric: MetricTimeSeries]) -> [NutritionCorrelation] {
        var results: [NutritionCorrelation] = []
        let calendar = Date.cal

        for pair in curatedPairs {
            guard let nutritionSeries = timeSeries[pair.nutrition],
                  let outcomeSeries = timeSeries[pair.outcome],
                  nutritionSeries.samples.count >= 14,
                  outcomeSeries.samples.count >= 14 else { continue }

            // Build date-indexed lookup
            let nutritionByDate = buildDateLookup(samples: nutritionSeries.sortedSamples)
            let outcomeByDate = buildDateLookup(samples: outcomeSeries.sortedSamples)

            // Collect paired observations (with day offset)
            var nutritionValues: [Double] = []
            var outcomeValues: [Double] = []

            for (date, nutritionValue) in nutritionByDate {
                let outcomeDate: Date
                if pair.dayOffset > 0 {
                    outcomeDate = calendar.date(byAdding: .day, value: pair.dayOffset, to: date) ?? date
                } else {
                    outcomeDate = date
                }

                if let outcomeValue = outcomeByDate[outcomeDate] {
                    nutritionValues.append(nutritionValue)
                    outcomeValues.append(outcomeValue)
                }
            }

            // Require >= 30 paired observations before claiming a nutrition–outcome
            // link. Below this, Pearson correlation is too noisy to ship as an
            // "X affects Y" insight to the user.
            guard nutritionValues.count >= 30 else { continue }

            // Compute Pearson correlation
            let r = pearsonCorrelation(x: nutritionValues, y: outcomeValues)
            guard abs(r) >= 0.15 else { continue }  // Skip weak correlations

            // Threshold detection: find where impact becomes pronounced
            let threshold = detectThreshold(nutritionValues: nutritionValues, outcomeValues: outcomeValues)

            // Generate actionable insight
            let insight = generateInsightText(
                pair: pair,
                correlation: r,
                threshold: threshold,
                sampleCount: nutritionValues.count
            )

            results.append(NutritionCorrelation(
                nutritionMetric: pair.nutrition,
                outcomeMetric: pair.outcome,
                correlation: r,
                timingEffect: nil,  // Simplified: full timing analysis needs sub-daily data
                thresholdValue: threshold?.value,
                actionableInsight: insight,
                sampleCount: nutritionValues.count
            ))
        }

        return results.sorted { abs($0.correlation) > abs($1.correlation) }
    }

    /// Generate insights from discovered correlations
    static func generateInsights(from correlations: [NutritionCorrelation]) -> [Insight] {
        correlations.prefix(5).map { corr in
            Insight(
                metric: corr.nutritionMetric,
                title: Copy.Analysis.NutritionCorrelation.affects(nutrition: corr.nutritionMetric.displayName, outcome: corr.outcomeMetric.displayName),
                summary: corr.actionableInsight,
                recommendation: Copy.Analysis.NutritionCorrelation.monitorToOptimize(nutrition: corr.nutritionMetric.displayName, outcome: corr.outcomeMetric.displayName),
                severity: abs(corr.correlation) > 0.4 ? .warning : .info,
                trend: .stable,
                currentValue: corr.correlation,
                baselineValue: 0,
                deviationPercent: abs(corr.correlation) * 100,
                category: .nutritionCorrelation,
                relatedMetrics: [corr.outcomeMetric],
                context: InsightContext(
                    confidenceLevel: min(1.0, Double(corr.sampleCount) / 60.0),
                    dataPointCount: corr.sampleCount
                )
            )
        }
    }

    // MARK: - Helpers

    private static func buildDateLookup(samples: [MetricSample]) -> [Date: Double] {
        let calendar = Date.cal
        var lookup: [Date: Double] = [:]
        for sample in samples {
            lookup[calendar.startOfDay(for: sample.date)] = sample.value
        }
        return lookup
    }

    private static func pearsonCorrelation(x: [Double], y: [Double]) -> Double {
        let n = Double(x.count)
        guard n > 2 else { return 0 }

        let sumX = x.reduce(0, +)
        let sumY = y.reduce(0, +)
        let sumXY = zip(x, y).reduce(0) { $0 + $1.0 * $1.1 }
        let sumX2 = x.reduce(0) { $0 + $1 * $1 }
        let sumY2 = y.reduce(0) { $0 + $1 * $1 }

        let numerator = n * sumXY - sumX * sumY
        let denominator = ((n * sumX2 - sumX * sumX) * (n * sumY2 - sumY * sumY)).squareRoot()

        guard denominator > 0 else { return 0 }
        return numerator / denominator
    }

    /// Find the threshold level where impact becomes pronounced using quartile analysis
    private static func detectThreshold(
        nutritionValues: [Double],
        outcomeValues: [Double]
    ) -> (value: Double, impactPercent: Double)? {
        let sorted = zip(nutritionValues, outcomeValues).sorted { $0.0 < $1.0 }
        let n = sorted.count
        guard n >= 20 else { return nil }

        // Split into quartiles
        let q3Start = n * 3 / 4
        let bottomHalfOutcome = sorted.prefix(n / 2).map(\.1)
        let topQuartileOutcome = sorted.suffix(from: q3Start).map(\.1)

        guard !bottomHalfOutcome.isEmpty, !topQuartileOutcome.isEmpty else { return nil }

        let bottomAvg = bottomHalfOutcome.reduce(0, +) / Double(bottomHalfOutcome.count)
        let topAvg = topQuartileOutcome.reduce(0, +) / Double(topQuartileOutcome.count)

        guard bottomAvg != 0 else { return nil }
        let impact = ((topAvg - bottomAvg) / abs(bottomAvg)) * 100

        guard abs(impact) >= 5 else { return nil }

        // Threshold is the value at the start of Q3
        let thresholdValue = sorted[q3Start].0

        return (value: thresholdValue, impactPercent: impact)
    }

    private static func generateInsightText(
        pair: CuratedPair,
        correlation: Double,
        threshold: (value: Double, impactPercent: Double)?,
        sampleCount: Int
    ) -> String {
        let direction = correlation > 0 ? "increases" : "decreases"
        let strength = abs(correlation) > 0.5 ? "strongly" : abs(correlation) > 0.3 ? "moderately" : "slightly"

        var text = "Your \(pair.nutrition.displayName) \(strength) \(direction) \(pair.outcome.displayName)"
        if pair.dayOffset > 0 {
            text += " the next day"
        }
        text += " (r=\(String(format: "%.2f", correlation)), n=\(sampleCount))."

        if let threshold {
            let formatted = pair.nutrition.formatValue(threshold.value)
            text += " Impact becomes more pronounced above \(formatted) \(pair.nutrition.unit) (\(String(format: "%.0f", abs(threshold.impactPercent)))% change)."
        }

        return text
    }
}

// MARK: - InsightAnalyzer Conformance

extension NutritionCorrelationAnalyzer: InsightAnalyzer {
    static var analyzerID: String { "nutritionCorrelation" }
    static var insightCategory: InsightCategory { .nutritionCorrelation }

    static func generateInsights(context: AnalysisContext) -> [Insight] {
        let correlations = analyze(timeSeries: context.timeSeries)
        return generateInsights(from: correlations)
    }
}
